diff -druN ../score/EigenLayrDelegation.sol core/EigenLayrDelegation.sol
--- ../score/EigenLayrDelegation.sol	2023-01-06 14:39:35
+++ core/EigenLayrDelegation.sol	2023-01-09 14:13:40
@@ -138,10 +138,10 @@
      */
     function decreaseDelegatedShares(
         address staker,
-        IInvestmentStrategy[] calldata strategies,
-        uint256[] calldata shares
+        IInvestmentStrategy[] memory strategies, // MUNGED calldata => memory
+        uint256[] memory shares // MUNGED calldata => memory
     )
-        external
+        public // MUNGED external => public
         onlyInvestmentManager
     {
         if (isDelegated(staker)) {
