diff -druN ../score/EigenLayrDelegation.sol core/EigenLayrDelegation.sol
--- ../score/EigenLayrDelegation.sol	2023-01-03 12:27:33.000000000 -0800
+++ core/EigenLayrDelegation.sol	2023-01-05 10:11:40.000000000 -0800
@@ -138,10 +138,10 @@
      */
     function decreaseDelegatedShares(
         address staker,
-        IInvestmentStrategy[] calldata strategies,
-        uint256[] calldata shares
+        IInvestmentStrategy[] memory strategies, // MUNGED calldata => memory
+        uint256[] memory shares // MUNGED calldata => memory
     )
-        external
+        public // MUNGED external => public
         onlyInvestmentManager
     {
         if (isDelegated(staker)) {
